### DISCUSSION TITLE
Hellfire content with Diablo systems?

### DISCUSSION THREAD

**Discussion by agris-codes** [diasurgical/DevilutionX#1804] [Ideas]
Hi All-

Longtime, as in _loooong time_ fan of D1 here, I recently became aware of this project and it's very impressive. I've been browsing the PRs, discussions and issue tracker to get a sense for the state of the project, as well as community priorities, and I wanted to share some feedback on Hellfire implementation.

I understand that Hellfire to be based upon Diablo v1.02, and to have implemented some itemization and systems changes that were never implemented in base Diablo 1. Things such as changing the starting staff of the Sorc, initial spell cost, and likely many many under the hood mechanics changes.

Snippet from [another forum](https://rpgcodex.net/forums/threads/diablo-1.97935/page-5#post-3851751):

> Hellfire completely [screws up initial Sorc difficulty] by switching out your starting staff of Charged Bolt (powerful but runs out fast) with a Staff of Mana (essentially like carrying an extra 18 mana potions every time you enter the labyrinth, and it recharges for like 300 gold where potions cost 50 each). For some reason beyond understanding they also cut his starting spell costs (normally spells start costing much more to cast then with higher levels they go down to about 50% as much, sorcerer starts at min costs). So Sorcerers in Hellfire can just throw spells around like candy, clearing out the first 3-4 floors of the labyrinth in one go, where normally you'd be limping back to town at least once a level, spending half your gold on more mana potions, and using bows to kill any slow zombies.

This has me wondering:

1. Are the systems-level changes that Hellfire instituted over vanilla Diablo (i.e. magic cost) documented in a comprehensive manner anywhere?
2. When devilutionX is only using the vanilla Diablo MPQ, not Hellfire's additional file, is it implementing Diablo's or Hellfire's systems and mechanics - such as the Sorc starting staff and spell costs
3. Assuming devilutionX is toggling between the two implementations of game mechanics depending on which MPQs are found, is it possible to (a) play Hellfire using Diablo's mechanics, or (b) play Diablo using Hellfire's mechanics?

I'm sure basic things like this were hashed out much earlier in the project, so feel free to reference earlier discussions. My goal here is to play through Hellfire with the core mechanics of Diablo v1.09b plus all the new options afforded by devilutionX.

Sorry for such a long first post!

PS - how is "hold LMB to continue attacking" not a feature?! I guess I need to fire up d2clicker.

**Comment by AJenbo** [comment]
hellfire is based on what is much closer to 1.04, except for one bugfix.

1 you can see all the hellfire differences by searching for gbIsHellfire https://github.com/diasurgical/devilutionX/search?q=gbIsHellfire

2 The logic is also toggeled

3 You can change `gbIsHellfire` to `0` for the parts of hellfire that you would like to avoid

**Reply by agris-codes** [reply]
Thanks @AJenbo, so if I could re-state your reply to (3), is it fair to say: there is no high-level toggle for Diablo-like or Hellfire-like mechanics, and the only way to get control over that is via the source modification you indicated?

No problem if so, just making sure I'm clear on it.

**Reply by AJenbo** [reply]
correct, some people have been thinking about making a mod like that, but its not something we plan on adding as we try not to change things that where deliberate choices by the original devs.

**Reply by agris-codes** [reply]
It's not too crazy of a modification, playing Hellfire with Diablo mechanics that is, if you consider that (1) Hellfire, while a licensed product, was not made by the same development team as Diablo and (2) during Blizzard's post-release support for Diablo 1, which I *believe* extended pass Hellfire's release on 11/25/1997, patches were issued for Diablo which did not elect to port over Hellfire's changes, such as spell cost. 

This may be a good intro to github project for myself, actually.

**Reply by AJenbo** [reply]
We have already applied the Diablo 1.04 - 1.09 patches to Hellfire.

**Reply by agris-codes** [reply]
> 
> 
> We have already applied the Diablo 1.04 - 1.09 patches to Hellfire.

Understood, my point being that the Diablo developers did not elect to implement some of the Hellfire mechanical changes in their post-Hellfire release Diablo patches. With that observation, playing Hellfire content with Diablo mechanics may actually be closer to the _Diablo_ devs intent, but obviously not that of the Hellfire team at Sierra.

**Reply by AJenbo** [reply]
Diablo 1.07 actually does contain mechanics that where added (slightly differently) in Hellfire 1.01, which came out first. Its not easy to say what they liked and what they didn't, as they didn't share patches and most development effort was going in to Diablo 2 at the time (that's why Sierra was contracted to do the expansion).

**Comment by agris-codes** [comment]
Now, I very much feel as if I am a caveman poking at fire, but I have gotten to the point where I can compile the master branch on my windows machine, and the resulting devilution.exe does not crash with limited testing, so I feel as if I am in a good place to try to externalize some of the hellfire vs diablo mechanics when Hellfire's MPQ is present.

Assume the default environment is Diablo + Hellfire MPQs, I plan to created .ini switches to do the following:

- Remove Hellfire maps
- Remove Hellfire quests
- Remove Hellfire character classes
- Remove Hellfire potions
- Remove Hellfire items
- Revert Sorc's starting staff to Diablo's default
- Revert Sorc spell cost scaling to Diablo's default

That gives players with both Hellfire & Diablo a good set of options to customize their game, without the barrier of entry that source modification + compilation poses. 

Other suggestions for options are welcome!

**Reply by julealgon** [reply]
When you say "I  plan to created .ini switches", do you mean in DevilutionX or in a mod?

If you mean in DevilutionX, then I disagree with all your entries as ini flags. Maps/quests/character classes/items/starting items should be softcoded and customized that way (not through .ini). As for the last point, I don't think we should have 2 distinct strategies for handling cost scaling so I don't think your last toggle option is valid.

**Reply by agris-codes** [reply]
@julealgon As a mod. I think it bears more discussion than a simple "I disagree" for inclusion in DevilutionX, but that's getting ahead of myself. I need to create a viable product before that discussion is relevant.

To your point of softcoding - as in through an in-game menu?

edit: also, I included your cancel attack PR in my build of the current master, so I plan on trying that change out in an actual game.

**Reply by AJenbo** [reply]
softcoding = in can be changed in a data file rather then by having to change code. For examples there could be a text file with a list of item types and you could simply delete Hellfire potions from it and start the game and it would no longer have hellfire potions.

**Reply by agris-codes** [reply]
For my specific set of proposed options for when Hellfire's MPQ is present, I don't see a functional difference between editing an external text file and editing a .ini file. There's quite a few customization switches already contained in the .ini file, so it isn't as if this is new territory.

Although, the ability the set these options from an in-game menu is intriguing. And likely beyond my immediately ability. But such a model is more friendly option for builds for mobile, where file system access and modification is more difficult or impossible.

**Reply by julealgon** [reply]
What I mean by "softcoding" is extracting the game data into modifiable files. For example, instead of having the 5 classes hardcoded in the game, the class definitions are extracted into a .json file, including all the related data for the class _and_ its starting items. 

Example:

```json
{
    "classes": [{
            "name": "Warrior"
            "portrait": "pathToTheHeroSelectionPortraitInTheMpqOrSomething",
            "attributes": {
                "strength": {
                    "base": 30,
                    "maximum": 250
                },
                "magic": {
                    "base": 10,
                    "maximum": 50,
                    "manaIncrement": 1
                },
                "dexterity": {
                    "base": 20,
                    "maximum": 60
                },
                "vitality": {
                    "base": 25,
                    "maximum": 100,
                    "lifeIncrement": 2
                }
            },
            "resources": {
                "life": {
                    "base": 70,
                    "bonusPerLevel": 2
                },
                "mana": {
                    "base": 10,
                    "bonusPerLevel": 1
                }
            },
            "traits": {
                "meleeToHitBonus": 30,
                "rangedToHitBonus": 10,
                "magicToHitBonus": 0,
                "blockingBonus": 30
            },
            "items": {
                "leftHand": "short sword",
                "rightHand": "buckler",
                "inventory": [
                    "club",
                    "100 gold"
                ],
                "belt": [
                    "health potion",
                    "health potion"
                ]
            },
            "skills": [{
                    "name": "Repair",
                    "iconImage": "pathToRepairImage"
                }
            ],
            "animations": {
                "townIdle": {
                    "frameData": {
                        "lightArmor": {
                            "oneHandedSword": {
                                "south": "pathToSAnimationOnMpq",
                                "southWest": "pathToSWAnimationOnMpq",
                                ...
                            },
                            "swordShield": {
                                ...
                            },
                            ...
                        },
                        "mediumArmor": {
                            ...
                        },
                        ...
                    }
                    "frameCount": 10,
                    "hitFrame": 5
                    "animationDelay": 2
                },
                ...
            },
            "sounds": {
                "overburdened": [
                    "pathToOverburdened1",
                    "pathToOverburdened2",
                    "pathToOverburdened3",
                ],
                "takeDamage": [
                    ...
                ]
            },
            "cinematics": {
                "ending": "pathToWarriorEnding"
            }
        }, {
            "name": "Rogue",
            ...
        },
        ...
    ]
}
```

Then, to remove a class, you just remove the node in the `classes.json` corresponding to the class you don't want.

Similarly, this would exist for items, for quests, for maps, etc.

**Reply by julealgon** [reply]
@sumyunggai 
> There's quite a few customization switches already contained in the .ini file, so it isn't as if this is new territory.

The difference is in the "type" of toggle. The toggles in .ini are generally not data-related toggles (the ones that are there today I disagree with, like the "allquests" one). For data in general, it makes more sense to have the data itself extracted as you can add infinite flexibility that way, instead of creating very specific toggles in the .ini.

I hope my example above made sense to you.

**Reply by agris-codes** [reply]
It does, and I can see the utility in having definitions and lists externalized to json format like you demonstrated.

That being said, there are some things that aren't well suited to such enumeration. From my bullet points, (a) disabling of hellfire's maps and (b) reverting the Sorc's spell cost changes. 

The ability to select "Diablo" or "Hellfire" behavior across key areas of the game is not directly relatable to your goal of infinite flexibility. What you are proposing is a good long-term solution to the mod-ability of devilutionX. What I'm proposing is very specific to the changes instituted to Diablo by the Hellfire expansion, and providing the user the ability to decide, when playing with the Hellfire MPQ, which of Sierra's changes they want to include in their game. 

Given Hellfire's reception and reputation, I think my goal is more mainstream then a generalized infinite flexibility of game data via JSON files. Gating the implementation of selecting between Diablo or Hellfire behavior across those areas of gameplay behind externalization of game data conflates two very different goals, to my mind. It also puts a higher barrier to implementation, and has a bit of a "the perfect being the enemy of the good" feel to it - with the conceit that I think my thesis of providing the general options outlined above is a good thing :)

That being said, would you or @AJenbo mind pointing me to source files dealing with the population and rendering of game options when ESC is hit? Despite AJembo not being keen on my idea of in-game menu options, this discussion has reinforced how helpful it would be if players could make such selections in-game. Rather than binary .ini toggles, it probably makes sense to offer 3 selection per bullet point: Default, Diablo, Hellfire - with default defaulting to the behavior based on which MPQs are installed.

**Reply by AJenbo** [reply]
? I'm very keen on an in-game option for editing option. Don't know how you would get the opposite impression. See: https://github.com/diasurgical/devilutionX/issues/1561 Basically uses should not know that the ini file exists unless they want to do something advanced that normal users shouldn't be exposed to either.

But an ingame option doesn't make much sens with out a ini option to save it to. Else you will have to change the setting each time you start the game.

**Reply by agris-codes** [reply]
Thanks @AJenbo, when I read your post below saying in-game and .ini were essentially the same thing, I was assuming you didn't like either. 

So all in-game menu item selections have to be tracked via an external file, even vanilla options like Gamma? That link looks great, could you point me to a source file or two that governs these menus?

I know that modern convention is that we have these kind of discussions on something like Discord, and apologies for not using it. Personally, I find the growth of platforms such as Discord commensurate with the loss of knowledge afforded by searchable forums. The chats are transitory, and don't provide any value to potential future players searching for information. That's why I'm making a point to use the Discussions section.

**Reply by AJenbo** [reply]
> So all in-game menu item selections have to be tracked via an external file, even vanilla options like Gamma?

How else would the application remember it? The original saved most of the settings (like gamma) in to the Windows registry, but that is both platform specific and not a very good system imo, and in the end the registry is saved to an external file as well.

> That link looks great, could you point me to a source file or two that governs these menus?

https://github.com/diasurgical/devilutionX/blob/master/Source/help.cpp
(I'm currently cleaning up this file, the code there isn't very reusable so probably you would have to write most of it from scratch)

> chats are transitory, and don't provide any value to potential future players searching for information. That's why I'm making a point to use the Discussions section.

Discord actually has a fairly decent search function for what it is worth. And it really shines when it comes to reducing noise like "How do I install VS" from appearing in every GitHub issue.

**Reply by agris-codes** [reply]
Thanks, and I'm trying to be conscious of not contributing to the noise. That's part of why I'm just now compiling from source 15 days after we starting talking about this :) But mostly it's that my responsibilities don't particularly care that I've found a competent remake of one of my favorite games from childhood...

**Reply by AJenbo** [reply]
GitHub Discussion ads a good middle ground, but feel free to come on the chat and ask some basic questions, or simply just hang out.

**Comment by AJenbo** [comment]
> There's quite a few customization switches already contained in the .ini file, so it isn't as if this is new territory.

Correction, there are already to many :D

The idea behind softcodeing is that you can easily add/delete/modefy individual items. With ini switches you can only enable or disable a predefined behavior.

> Although, the ability the set these options from an in-game menu is intriguing. And likely beyond my immediately ability. But such a model is more friendly option for builds for mobile, where file system access and modification is more difficult or impossible.

An ingame menu would be editing the ini file so there isn't really a distinction, just a nicer user interface for doing the same thing.

**Comment by qqkookie** [comment]
I played Hellfire for a while. Please, consider my experience on Hellfire for selecting the group of feature set to make optional.

I like:
- Search spell: it should be introduced as press "alt" to show items on floor and map. Real saver for my aging eye to reduce eye fatigue and strain.  
- Warp spell: Better than Teleport or Phasing
- Nova spell book (because Flash spell is useless)
- New Unique items (esp. amulet/rings)
- Auric Amulet (gold 5K -> 10K limit to buy expensive (>200K)  item from  Griswold
- New unique mob (group leader) with distinct color pallet in a mob group
- Item transfer to other single character through crypt tile. 
- Speedwalk in town

Not bad, don't care:
- Monk class,  it is optional anyway.
- New maps and new quests, it is optional anyway
- New Shrines (can be avoided)
- Oils. As long as temporary boost, not permanent.

I dislike or think useless:
- New spells:  Berserk, Ring of Fire, Immolation, Lightning Wall, Reflect
- Apocalypses spell book
- The rune items (trap),
- New item modifiers (affix/suffix) (like jester's, doppelganger's)
- I hate Griswold trying to sell better items than I have.  I need +30 strength Titan ring or +30 magic Wizard ring though I have +20 all stats Obsidian Ring of Zodiac. I like it to be half-and-half.
- I agree that the sorcerer's starting staff is too much. 

For feature set selection, we can use the existing option mechanism in `Diablo.ini`: DevilutionX already has few Hellfire related options like  `"Theo Quest,Cow Quest,Test Bard,Test Barbarian"` in `Source/option.h/struct GameplayOptions`, why not add more?

And please consider the "hold to shoot" (hold LMB to shoot continuously) feature. My aging carpal tunnel is suffering.

**Reply by julealgon** [reply]
I hope you realize that, if we were to introduce flags for every "likes" and "dislikes" of every single player out there, we'd create a toggle system and not a game. We cannot add a toggle for every possible behavior ever otherwise this will become a maintenance hell.

> DevilutionX already has few Hellfire related options like "Theo Quest,Cow Quest,Test Bard,Test Barbarian" in Source/option.h/struct GameplayOptions, why not add more?

I personally don't think any of those should've been added as they were. So justifying any new toggles based on those is a moot point in my opinion.

> And please consider the "hold to shoot" (hold LMB to shoot continuously) feature. My aging carpal tunnel is suffering.

May I ask what does that have to do with the current topic?

**Reply by AJenbo** [reply]
> > DevilutionX already has few Hellfire related options like "Theo Quest,Cow Quest,Test Bard,Test Barbarian" in Source/option.h/struct GameplayOptions, why not add more?
> 
> I personally don't think any of those should've been added as they were. So justifying any new toggles based on those is a moot point in my opinion.

Thease are original options from Hellfire, they where simply ported over. Why not add more runes?

**Reply by qqkookie** [reply]
> I hope you realize that, if we were to introduce flags for every "likes" and "dislikes" of every single player out there, we'd create a toggle system and not a game. We cannot add a toggle for every possible behavior ever otherwise this will become a maintenance hell.

There are  already nitty-gitty detailed options like these. IMHO, the hellfire likes-dislikes are more meaningful then these. 
```
Auto Equip Sound=0
Auto Equip Weapons=1
Auto Equip Armor=0
Auto Equip Helms=0
Auto Equip Shields=0
Auto Equip Jewelry=0
```

> > And please consider the "hold to shoot" (hold LMB to shoot continuously) feature. My aging carpal tunnel is suffering.
>
> May I ask what does that have to do with the current topic?

Author said it in PS of opening post.
> PS - how is "hold LMB to continue attacking" not a feature?! I guess I need to fire up d2clicker.

> > > DevilutionX already has few Hellfire related options like "Theo Quest,Cow Quest,Test Bard,Test Barbarian" in Source/option.h/struct GameplayOptions, why not add more?
> > 
> > 
> > I personally don't think any of those should've been added as they were. So justifying any new toggles based on those is a moot point in my opinion.
> 
> Thease are original options from Hellfire, they where simply ported over. Why not add more runes?

So how about move these Hellfire specific flags to `struct HellfireOptions`?

I think that the `options.h` and format of `Diablo.ini` should be clean up for more sane and user-friendly fashion.

Data structure in `options.h` should be adjusted or renamed accordingly to make it clear that they are devilution specific improvements. ex) `struct GamePlayOptions -> struct DevXOptions`.

And I also suggest to add new [Mods] section and corresponding `struct ModsOptions` structure stub to absorb proliferation of devilutionX-derived mods, not merged into Devilution Master branch, and their options to prevent adding new flag member in `struct GamePlayOptions (struct DeviXOptions)`. It is just stub for Mods. Master branch will not use it. 

Suggested `Diablo.ini` sections. Textual order of sections are adjusted.
```
; Diablo.ini
; To change location use --config=AltConfig.ini command line option
;
; Section for Diablo/Hellfire common global options or command line options
[Global]
Intro=0             ; both diablo and hellfire intro,   -n command line option
Mpq Dir=.           ; --data-dir command line option,  data dir is too ambiguous. 
Save Dir=./SAVE     ; --save-dir
Language=en     ;  was [Langguage] Code=en
Font=./CharisSILB.ttf  ; --ttf-dir & ttf-name
Grab Input=0
Speed=20
Run in Town=0

; Section for Hellfire specific options
[Hellfire]
Force Diablo=1      ; --diablo command line option
Theo Quest=0
Cow Quest=0
Test Bard=0
Test Barbarian=0
Alt NestArt=0       ; --nestart command line option
SItem=

; Here goes additional Hellfire options
Ext Spells=1
Ext Items=1
Soc Staff=1

[Audio]
.......

[Graphics]
.....
Fullscreen=0   ; -x command line option
Show FPS=0     ; -f command line option

; Section for Devilution features and QOL improvements. 
[Options]
Randomize Quests=1
Friendly Fire=1
Disable Crippling Shrines=0
Auto Gold Pickup=0
Adria Refills Mana=0
Show Monster Type=0
Experience Bar=0
Enemy Health Bar=0
Auto Equip Weapons=1
Auto Equip Armor=0
Auto Equip Helms=0
Auto Equip Shields=0
Auto Equip Jewelry=0

; Section for option of Deviltion derived mods
[Mods]


; Sections for Keymaps and controller
[Keymapping]
Help=F1
ItemInfo=
QuestDebug=
QuickSpell1=F5
.....
CheatExperience=
StopHero=

[Controller]
Mapping=
Swap Shoulder Button Mode=0
Dpad Hotkeys=0
deadzone=0.070000

; Sections for multiplayer play
[Network]
Bind Address=0.0.0.0
Port=6112
Previous Host=

[NetMsg]
F9=I need help! Come Here!
F10=Follow me.
F11=Here's something for you.
F12=Now you DIE!
QuickMessage1=I need help! Come Here!
QuickMessage2=Follow me.
QuickMessage3=Here's something for you.
QuickMessage4=Now you DIE!
```

**Reply by AJenbo** [reply]
Diablo.ini is not intended to be user friendly. See https://github.com/diasurgical/devilutionX/issues/1561

**Reply by qqkookie** [reply]
> Diablo.ini is not intended to be user friendly. See #1561

Oh, what a wonderful feature! I am anticipating to see it released soon. But cleaning up the organization of `Diablo.ini` (and `options.h struct's`) will help to make the advanced option menu look more organized and better to navigate in both user view and source code. I think that the `Diablo.ini` organization will be better when it reflects the on-screen menu view organization (section and order) of the such advanced option menu screen.  So both intentions are not conflicting objectives. Both can help each other. Some entries not intended to be modified by player can be ignored/skipped  by advanced option menu.

**Reply by julealgon** [reply]
>There are already nitty-gitty detailed options like these. IMHO, the hellfire likes-dislikes are more meaningful then these.
>
> Auto Equip Sound=0
Auto Equip Weapons=1
Auto Equip Armor=0
Auto Equip Helms=0
Auto Equip Shields=0
Auto Equip Jewelry=0

Those options are presented that way for 2 (IMHO good) reasons:
1. Our ini library and C++ in general handles enums poorly. This would be much simpler as an enum-flags setting. Say:
`Auto Equip = Weapons, Armor, Shields`

2. The fact that there is a flag for each possible equipment type simplifies the code and prevents special-casing logic.

Those 2 are very technical concerns, as you can see.

I'm not against "detailed" options, I'm against "specific" ones. Those are different. While the settings above are detailed, they provide a very general system to toggle all the behaviors and is thus not "opinionated" in its implementation. This is why I dislike having only a few quests togglable in the INI: either we should be able to toggle everything, or nothing. That's my view from a code and architecture simplification standpoint.

> So how about move these Hellfire specific flags to struct HellfireOptions?

There should NOT be any "hellfire-specific flag" in my opinion. The flags should be tied to the behaviors, not whether or not they originally come from Diablo or Hellfire, that's not relevant and leads into special-casing in code, which is undesirable.

> Data structure in options.h should be adjusted or renamed accordingly to make it clear that they are devilution specific improvements. ex)

Same as above. I do not like the idea of "differentiating" between "devilution" things, "diablo" things and "hellfire" things. The originator by itself is meaningless: it does not matter. What matters is the functionality.

DevilutionX is supposed to be a _new_ thing, that is a combination of all aspects. Trying to perpetuate these distinctions will just lead to a more convoluted codebase and I'd rather we don't go that route.

> And I also suggest to add new [Mods] section and corresponding struct ModsOptions structure stub to absorb proliferation of devilutionX-derived mods, not merged into Devilution Master branch, and their options to prevent adding new flag member in struct GamePlayOptions (struct DeviXOptions). It is just stub for Mods. Master branch will not use it.

That's for a mod to decide IMHO, not DevilutionX. While DevilutionX is supposed to be a base for mods, it shouldn't be as opinionated as _that_ to me. It provides a base: if a mod wants to have its settings in a dedicated "[Mods]" section, it's up to that mod. Most mods will actually probably restrict settings, not make them even more open.

Keep in mind that DevilutionX currently doesn't have any "plug-in like" mod system like other more developed games have. When we say "mod support", at least for now, we mean "modifying the code directly". In effect, we actually _do not support mods_ at this stage. The difference between vanilla and DevilutionX is that you have access to the code in DevilutionX but not in vanilla: that's it.

**Reply by qqkookie** [reply]
> Those options are presented that way for 2 (IMHO good) reasons:
> Those 2 are very technical concerns, as you can see.

Agreed. C++ enum  support is lukewarm at best.
 
> I'm not against "detailed" options, I'm against "specific" ones.

I wholeheartedly agree. What I suggested is not to implement like "RemoveHellfireSpells=0", "RemoveHellfireItems=0" in such way. I agree that it should be implemented like "RemoveHellfireSpells= Berserk, Ring of Fire, Immolation, Lightning Wall, Reflect"

> > So how about move these Hellfire specific flags to struct HellfireOptions?
> 
> There should NOT be any "hellfire-specific flag" in my opinion.

> I do not like the idea of "differentiating" between "devilution" things, "diablo" things and "hellfire" things. The originator by itself is meaningless: it does not matter. What matters is the functionality. 

> DevilutionX is supposed to be a _new_ thing, that is a combination of all aspects. Trying to perpetuate these distinctions will just lead to a more convoluted codebase and I'd rather we don't go that route.

Is DeviltionX going to pursue route like _Belzebub HD_ MOD or _TheHell_ MOD in future, in just different way? My initial (and maybe incorrect) impression on playing DevilutionX was that it closely adheres to the old tradition (no hold to shoot!)  of Diablo/Hellfire originals, kinda a traditionalist or purist branch of Diablo MOD scene.  I am new to here so forgive my ignorance. 

Hellfire is not standalone expansion, but expansion pack to Diablo 1. The weight of diablo original versus Hellfire is so heavy.   And distinction between Diablo and Hellfire is so deeply rooted in the Diablo 1 tradition and fan base. Hellfire is treated like a red-haired stepchild by fan. The very existence of this discussion thread (Hellfire contents on Diablo system) proves it home. Nobody askes _Diablo 2:LOD_ contents in _Diablo 2_.  Such distinction is outright obvious in they way the Hellfire is implemented in source code.  I think the that such distinction will not go way in anytime soon. So I did not expect the Devilution become **something new** like Belzebub or TheHell. 

Same can be said to the DevilutionX. Fan will always recognize it as "Faithful Diablo re-implementation with some improvements", not something new, not even as much  new as Belzebub, TheHell, not to mention of Hellfire is new to Diablo. In source code navigation, what is new or different from original is vey important clue for  selecting where to look closely. If something is broken or not working as expected, then the root source of the functionality is very helpful hint to identify what is culprit.  So Hellfire/devilution related globals and the devolution-specific option flags in the `struct GamePlayOptions` are very important clue to identify the difference in source of functionality.  See  quote below:
> 1 you can see all the hellfire differences by searching for gbIsHellfire https://github.com/diasurgical/devilutionX/search?q=gbIsHellfire

> That's for a mod to decide IMHO, not DevilutionX. While DevilutionX is supposed to be a base for mods, it shouldn't be as opinionated as _that_ to me. It provides a base: if a mod wants to have its settings in a dedicated "[Mods]" section, it's up to that mod. Most mods will actually probably restrict settings, not make them even more open.

This will lead to proliferation of new member of `struct GamePlayOptions` which is already heavily burdened and `Diablo.ini` sections with just one or two entries. The suggestion is not enforcing use of struct ModsOptions and [Mods], but it softly nudges or give hint to modders to avoid adding new members and section.  As I said above, source of the functionality at hand is very helpful hint to locate where to look at. 

> Keep in mind that DevilutionX currently doesn't have any "plug-in like" mod system like other more developed games have. When we say "mod support", at least for now, we mean "modifying the code directly". In effect, we actually _do not support mods_ at this stage. The difference between vanilla and DevilutionX is that you have access to the code in DevilutionX but not in vanilla: that's it.

Integrating Lua interpreter module to DevolutionX will completely change such current situation of  "no non-C++-developer-level MOD support". Mod in Lua code will come sooner or later, I expect. I guess that The Belzebub MOD has such Lua support.

**Reply by qndel** [reply]
" it closely adheres to the old tradition (no hold to shoot!)" - that's not our goal, atm "hold to attack" is not present more because of technical limitations than anything else

"So I did not expect the Devilution become something new like Belzebub"
DevilutionX will never go that way, however it's supposed to be a mod base, so definitely expect a mod based on devilutionx that will do just that

**Reply by qqkookie** [reply]
> " it closely adheres to the old tradition (no hold to shoot!)" - that's not our goal, atm "hold to attack" is not present more because of technical limitations than anything else
> 
> "So I did not expect the Devilution become something new like Belzebub"
> DevilutionX will never go that way, however it's supposed to be a mod base, so definitely expect a mod based on devilutionx that will do just that

Happy to hear that. I am anticipating such day will come sooner than later. Thank you for kind explanation.

**Reply by Marcelo20XX** [reply]
So, are any of the ideas been implemented by the OP? I would assume on a separate branch since DevilutionX is more of a base for mods...

I would like to play with the Monk with Diablo system nothing else, no new spells, books, shrines or items, just that class added to the rooster of heroes...

**Reply by qndel** [reply]
No.

**Reply by agris-codes** [reply]
> So, are any of the ideas been implemented by the OP? I would assume on a separate branch since DevilutionX is more of a base for mods...
> 
> I would like to play with the Monk with Diablo system nothing else, no new spells, books, shrines or items, just that class added to the rooster of heroes...

Unfortunately not. A couple of things work against me- my programming background is almost orthogonal to CPP, I have very little time, the parameters I want to change are hardcoded in various ways and littered throughout source, and there was (as of 2021) no workable software structure to, in essence, toggle over various mechanics to Diablo or HF variants without restructuring many parts of the source to be more modular.

All that is to say, it is a task worth doing - in my opinion - but I lack the time, ability, and quite frankly the drive to make it happen.

**Reply by Marcelo20XX** [reply]
Where are you failing is in that you want to integrate the changes into the main branch of DevilutionX and that is not really needed, you can just branch your way and make the needed changes without worrying anyway, since its open source, other people can clean or incorporate your changes if they need/want to do so...

Not to worry though I am working on it already, I am just finished on a Minimal UI mod I was working and this will be my next project, I asked because I didn't want to duplicate efforts...

I will release any changes as a diff file if anyone is interested...

**Reply by agris-codes** [reply]
Great news! Hope you can parse the Diablo vs Hellfire mechanics into understandable chunks for players to tweak. I tried to outline those in my posts above.

**Reply by AJenbo** [reply]
Most people don't k now what to do with a diff file, probably just push it as a fork on your github account.

**Reply by Marcelo20XX** [reply]
My goal is different, I dislike Hellfire aside from the Monk Class, so I would probably only include that class to the rooster and nothing else...

EDIT: Well I will try to branch the repo then Ajenbo... with my minimal Ui mod included as a separate branch too and excuse me the mess of my code in advance, I usually only code to get things done, that's why I almost never attempt to make any PR...

**Reply by AJenbo** [reply]
branches still work

**Reply by agris-codes** [reply]
Every brick, no matter it’s purpose, contributes to the road. 

Your limited scope mod may inspire others with a broader vision, closer to my own but with more ability. I wish you luck!

**Reply by Marcelo20XX** [reply]
I added the Monk class to the character rooster of Diablo, if you want to try it here is the link as promised:
https://github.com/Marcelo20XX/devilutionX/commit/1a2b115db6d60d6656417dafc76bea426e5b8587
